### PR TITLE
Support sequence parallelism and collective matmul

### PR DESCRIPTION
**Pull request by yaochengji**
# Description

- add sequence parallelism sharding annotation for torchax dense models
- fix the issue of v6e-4, v6e-8 logical device id order
- the feature can be enabled by `-O '{"pass_config": {"enable_sequence_parallelism": true}}'`

# Tests

```
MODEL_IMPL_TYPE=vllm vllm serve Qwen/Qwen2.5-32B  --seed 42  --disable-log-requests  --tensor-parallel-size 8 --max-model-len 2048 --gpu-memory-utilization 0.96 --no-enable-prefix-caching --max-num-seqs 256 --max-num-batched-tokens 4096 -O '{"pass_config": {"enable_sequence_parallelism": true}}' |& tee run.log
```
```
python3 ./benchmarks/benchmark_serving.py --model Qwen/Qwen2.5-32B --dataset-name sonnet --dataset-path benchmarks/sonnet_4x.txt --sonnet-input-len 1800 --sonnet-output-len 128 --ignore_eos
```

```
MODEL_IMPL_TYPE=vllm vllm serve meta-llama/Llama-3.1-70B-Instruct  --seed 42  --disable-log-requests  --tensor-parallel-size 8 --max-model-len 2048 --gpu-memory-utilization 0.96 --no-enable-prefix-caching --max-num-seqs 128 --max-num-batched-tokens 2048 -O '{"pass_config": {"enable_sequence_parallelism": true}}' |& tee run.log
```

```
python3 ./benchmarks/benchmark_serving.py --model meta-llama/Llama-3.1-70B-Instruct --dataset-name sonnet --dataset-path benchmarks/sonnet_4x.txt --sonnet-input-len 1800 --sonnet-output-len 128 --ignore_eos
```

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

**Comment by yaochengji** [comment]
| Model    | Best_throuput_before | Throuput_new| Improvement|
| -------- | ----------------------|----------------|-------------|
|  Qwen/Qwen2.5-32B| 12.1req/s    |14.4req/s| 18.9%|
|  Llama-3.1-70B| 7.58req/s     |8.1req/s|6.9%|

**Comment by yaochengji** [comment]
cc @hfan could you also take a look?

cc @xiangxu-google , you may also apply this on jax models.

File: tpu_commons/utils.py [hunk 1]
```diff
@@ -106,3 +112,65 @@ def get_padded_num_heads(num_heads: int, sharding_size: int) -> int:
...
+                    devices[4],
+                    devices[6],
+                    devices[7],
+                    devices[5],
+                    devices[3],
+                    devices[1],
+                ])
+                ordered_devices = np.array(ordered_devices)
+                ordered_devices = ordered_devices.reshape(axis_shapes)
+                mesh = mesh_lib.Mesh(ordered_devices, axis_names)
```
Comment (hfan) [inline tpu_commons/utils.py]: The above 3 lines are repeated in the `if ordered_devices is not None` branch

Comment (yaochengji) [reply tpu_commons/utils.py]: Thanks for catching this!

**Comment by hfan** [comment]
@kyuyeunk - this affects your refactor too?

File: tpu_commons/models/vllm/vllm_model_wrapper.py [hunk 2]
```diff
@@ -138,6 +138,12 @@ def jit_step_func(self):
         @functools.partial(
             jax.jit,
             donate_argnums=(1, ),  # donate kv_cache
+            compiler_options={
+                "xla_tpu_all_gather_collective_matmul_mode":
+                "post_spmd_conservative",
+                "xla_tpu_reduce_scatter_collective_matmul_mode":
+                "post_spmd_conservative"
+            },
         )
```
Comment (hfan) [inline tpu_commons/models/vllm/vllm_model_wrapper.py:147]: OOC, if we don't fix the device order, do the collective matmul flags work? I don't recall if I've tried these flag, it could be because the device order issue that they didn't work.

File: tpu_commons/models/vllm/jax_row_parallel_linear.py [hunk 3]
```diff
@@ -96,4 +99,10 @@ def forward(self, input: torch.Tensor):
             if not self.return_bias:
                 return output
             output_bias = self.bias if self.skip_bias_add else None
+            if self.enable_sequence_parallelism:
+                token_num = input.shape[0]
+                # NOTE(chengjiyao): make sure the sharded token_num is larger than TPU_SECOND_LAST_MINOR
+                if token_num // self.mesh.shape[
+                        'model'] >= TPU_SECOND_LAST_MINOR:
+                    output.shard_(NamedSharding(self.mesh, P('model', None)))
```
Comment (hfan) [inline tpu_commons/models/vllm/jax_row_parallel_linear.py:107]: does it require the output to be reduced first? IOW, the `self.reduce_results` has to be true?

File: tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py [hunk 4]
```diff
@@ -230,7 +232,14 @@ def forward_split(self, input):
...
+                if token_num // self.mesh.shape[
+                        'model'] >= TPU_SECOND_LAST_MINOR:
+                    input.shard_(NamedSharding(self.mesh, P('model', None)))
             if self.fuse_matmuls:
-                return self.forward_fused(input)
+                output, output_bias = self.forward_fused(input)
             else:
-                return self.forward_split(input)
+                output, output_bias = self.forward_split(input)
+            return output, output_bias
```
Comment (hfan) [inline tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py:245]: so the output is sharded by P('model', None)? Does it just work for the later layers if the output is changed to be sharded this way?

File: tpu_commons/models/vllm/vllm_model_wrapper.py (hunk 2 above)
Comment (yaochengji) [reply tpu_commons/models/vllm/vllm_model_wrapper.py:147]: It works but the performance is much worse.

File: tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py [hunk 5]
```diff
@@ -230,7 +232,14 @@ def forward_split(self, input):
 
     def forward(self, input: torch.Tensor):
         with jax.named_scope(self.name):
+            if self.enable_sequence_parallelism:
```
Comment (hfan) [inline tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py:235]: Unit test probably is needed

**Comment by hfan** [comment]
Anything prevent it being turned on by default?

File: tpu_commons/models/vllm/jax_row_parallel_linear.py (hunk 3 above)
Comment (yaochengji) [reply tpu_commons/models/vllm/jax_row_parallel_linear.py:107]: From the definition of `reduce_results` is to add an `all_reduce` on the output of row_linear, but sequence_parallelism/collective_matmul requires it's a `reduce_scatter` instead.

File: tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py (hunk 5 above)
Comment (yaochengji) [reply tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py:235]: I intended to add an e2e test, but cannot find a good example.

File: tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py (hunk 4 above)
Comment (yaochengji) [reply tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py:245]: Emmm... what do you mean by `later layers`?

**Comment by yaochengji** [comment]
> Anything prevent it being turned on by default?

@hfan I'm not sure if this can benefit all the models, maybe @QiliangCui can first try this in auto-tuning

File: tpu_commons/models/vllm/jax_row_parallel_linear.py (hunk 3 above)
Comment (hfan) [reply tpu_commons/models/vllm/jax_row_parallel_linear.py:107]: I see, makes sense

File: tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py (hunk 5 above)
Comment (hfan) [reply tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py:235]: `tests/models/vllm/test_jax_XXX_linear.py` are the existing unit tests.

File: tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py (hunk 4 above)
Comment (hfan) [reply tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py:245]: nvmd, resolved in the offline discussion, thanks!

**Comment by kyuyeunk** [comment]
> @kyuyeunk - this affects your refactor too?

Yes, this does get affected. @yaochengji, I have created a draft PR so you can take an early look: https://github.com/vllm-project/tpu_commons/pull/512

**Comment by hfan** [comment]
> > Anything prevent it being turned on by default?
> 
> @hfan I'm not sure if this can benefit all the models, maybe @QiliangCui can first try this in auto-tuning

As discussed offline, quantized matmuls that uses kernels whose collectives can't be automatically handled by XLA, so we probably have to wait for your collective-matmul kernel and better be safe (or only make it default for the non-quantized code path)

File: tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py (hunk 5 above)
Comment (yaochengji) [reply tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py:235]: Thanks, @hfan ! The unit test is added, manually triggered in https://buildkite.com/tpu-commons/tpu-commons-ci/builds/2043

File: tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py [hunk 6]
```diff
@@ -230,7 +232,14 @@ def forward_split(self, input):
...
+                token_num = input.shape[0]
+                # NOTE(chengjiyao): make sure the sharded token_num is larger than TPU_SECOND_LAST_MINOR
```
Comment (vanbasten23) [inline tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py:237]: If "sharded token_num is larger than TPU_SECOND_LAST_MINOR", I guess the downside is waste of memory?

File: tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py [hunk 7]
```diff
@@ -230,7 +232,14 @@ def forward_split(self, input):
...
+                if token_num // self.mesh.shape[
+                        'model'] >= TPU_SECOND_LAST_MINOR:
+                    input.shard_(NamedSharding(self.mesh, P('model', None)))
```
Comment (vanbasten23) [inline tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py:240]: very neat and convenient way of doing `self.apply_jax_(jax.lax.with_sharding_constraint, sharding)`

File: tpu_commons/utils.py [hunk 8]
```diff
@@ -106,3 +112,62 @@ def get_padded_num_heads(num_heads: int, sharding_size: int) -> int:
...
+            # (0,1,0)
+            # (1,1,0)
+            # (0,2,0)
+            # (1,2,0)
+            # (0,3,0)
+            # (1,3,0)
+            if device_num == 8:
+                ordered_devices = np.array([
+                    devices[0],
+                    devices[2],
```
Comment (vanbasten23) [inline tpu_commons/utils.py:145]: I wonder why `(1,0,0)` maps to device `[2]`.

Also, does the order
```
            # (0,0,0)
            # (1,0,0)
            # (0,1,0)
            # (1,1,0)
            # (0,2,0)
            # (1,2,0)
            # (0,3,0)
            # (1,3,0)
```
matter?

Comment (yaochengji) [reply tpu_commons/utils.py:145]: Yeah, from the order, it matters

File: tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py (hunk 6 above)
Comment (yaochengji) [reply tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py:237]: There will be more communication in the final result, you can have a try.

File: tpu_commons/utils.py (hunk 8 above)
Comment (vanbasten23) [reply tpu_commons/utils.py:145]: why does `(1,0,0)` maps to device `[2]`?

File: tests/models/vllm/test_jax_merged_column_parallel_linear.py [hunk 9]
```diff
@@ -78,7 +80,10 @@ def test_jax_merged_column_parallel_linear(bias, mesh, fuse_matmuls):
     # Set jax default device to workaround a layout bug in JAX 0.7.0 and earlier
     with torchax.default_env(), jax.default_device(jax.devices("tpu")[0]):
```
Comment (vanbasten23) [inline tests/models/vllm/test_jax_merged_column_parallel_linear.py:81]: is it using one TPU device? Do you need to test when there are multiple TPU device?

File: tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py (hunk 7 above)
Comment (vanbasten23) [inline tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py:240]: It seems that SP is implemented by sharding the num_tokens dimension. Do you need to do an all-gather at the very end? I couldn't find it in your pr.

Comment (yaochengji) [reply tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py:240]: sharding propagation can handle this and make it correct.

Comment (vanbasten23) [reply tpu_commons/models/vllm/jax_merged_column_parallel_linear_core.py:240]: chatted offline, all-gather is not needed because some later ops (e.g. select which token to get logit (need to select num_reqs tokens)) may hint the compiler to get the global view. At that time, compiler will do a all-gather implicitly.

File: tpu_commons/models/vllm/jax_row_parallel_linear.py (hunk 3 above)
Comment (kyuyeunk) [inline tpu_commons/models/vllm/jax_row_parallel_linear.py:107]: Doesn't sequence parallelism usually use `'data'` to shard batch dim? Meaning, we can use both sequence and model parallelism and shard the inputs/outputs using `P('data', 'model')`

Comment (yaochengji) [reply tpu_commons/models/vllm/jax_row_parallel_linear.py:107]: In this case, only the mesh "model" axis is enough. Here the sequence parallelism is applied on layer_norm, and model parallelism is applied on matmul. It is described in this paper: https://arxiv.org/abs/2205.05198
